### PR TITLE
glib: add bin output for gvfs executables

### PR DESCRIPTION
**Pull request by Mic92**
this fixes gvfs in caja (mate desktop)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dezgeg** [comment]
Note that there is a related FIXME comment in someplace else: https://github.com/Mic92/nixpkgs/blob/ef73fa9c095eee8a337c172998e2fd6719100ff1/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L32

Does this perhaps fix that issue as well? If so, the no longer correct `${glib.dev}/bin` should be fixed in that file.

**Comment by Mic92** [comment]
Good point. Maybe gvfs binaries also should be patched to always have `gio` in `$PATH`

**Comment by jtojnar** [comment]
@dezgeg IIRC, the comment refers to `gsettings` program.

----

To put the issue into context, glib has three types of binaries:

- Ones needed during builds like `glib-mkenums`
- Ones needed after installation like `glib-compile-schemas`
- System utilities `gio` or `gsettings`

Optimally, they would be in separate packages but since glib is monolithic, we should use different outputs.

File: pkgs/development/libraries/glib/default.nix [hunk 1]
```diff
@@ -106,12 +106,15 @@ stdenv.mkDerivation rec {
   DETERMINISTIC_BUILD = 1;
 
   postInstall = ''
+    # needed by gvfs binaries
+    mkdir -p $out/bin
+    mv $dev/bin/gio $out/bin
```
Comment (jtojnar) [inline pkgs/development/libraries/glib/default.nix]: I think bin output would be a best place for system utilities like `gio` and `gsettings`.

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/glib/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/development/libraries/glib/default.nix]: @jtojnar can you help with what is needed when?

/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gio
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gio-querymodules
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gapplication
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gdbus
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/glib-compile-resources
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/glib-mkenums
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gdbus-codegen
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gobject-query
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gresource
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gtester
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gsettings
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/glib-gettextize
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/glib-compile-schemas
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/glib-genmarshal
/nix/store/zylnjb0wljm7abn96jk5jv59lj1lmb7l-glib-2.56.0-dev/bin/gtester-report

**Comment by jtojnar** [comment]
* `gapplication` – used to start D-Bus-activatable applications, sometimes used in `Exec` keys in destktop files as a fallback (see description of [DBusActivatable](https://specifications.freedesktop.org/desktop-entry-spec/1.2/ar01s06.html)), maybe it should be always be in user environment?
* `gdbus` – used for running some services (in my `/nix/store` I found it in `xdg-utils` which use it inside Flatpak and in `gnome3.tracker`, the latter should probably be patched), maybe some people may want to use it access D-Bus from shell scripts
* `gio` – replaces `gvfs-*` commands, some users might want to install them either for scripts or manual usage
* `gdbus-codegen`, `glib-compile-resources`, `glib-genmarshal`, `glib-mkenums` – used during build
* `glib-gettextize` – developers use it for adding gettext support to autotools projects? is it even necessary with gettext 0.19, glib gettext m4 macros are definitely discouraged in favour of uptsream ones?
* `gtester`, `gtester-report` – used for running tests (during build? by some autotools macro? meson has its own test runner)
* `gio-querymodules`, `glib-compile-schemas` – usually run after installation to [rebuild caches](https://github.com/NixOS/nixpkgs/issues/1455)
* `gobject-query` – no idea
* `gresource` – debugging `gresource` files created `glib-compile-resources`?
* `gsettings` – editing of GSettings from CLI, some users might want it even if it [does not work very well](https://github.com/NixOS/nixpkgs/issues/33277) with Nix

I think `gapplication`, `gdbus`, `gio`, `gsettings` and maybe `gresource` should go to `bin`, the rest to `dev`.

File: pkgs/development/libraries/gtkd/default.nix [hunk 2]
```diff
@@ -63,15 +63,15 @@ stdenv.mkDerivation rec {
       --replace libgtksourceview-3.0.so.1 ${gtksourceview}/lib/libgtksourceview-3.0.so.1 \
       --replace libgtksourceview-3.0.1.dylib ${gtksourceview}/lib/libgtksourceview-3.0.1.dylib
     substituteInPlace generated/gtkd/gtkc/glib.d \
-      --replace libglib-2.0.so.0 ${glib}/lib/libglib-2.0.so.0 \
-      --replace libgmodule-2.0.so.0 ${glib}/lib/libgmodule-2.0.so.0 \
-      --replace libgobject-2.0.so.0 ${glib}/lib/libgobject-2.0.so.0 \
-      --replace libglib-2.0.0.dylib ${glib}/lib/libglib-2.0.0.dylib \
-      --replace libgmodule-2.0.0.dylib ${glib}/lib/libgmodule-2.0.0.dylib \
-      --replace libgobject-2.0.0.dylib ${glib}/lib/libgobject-2.0.0.dylib
+      --replace libglib-2.0.so.0 ${glib.out}/lib/libglib-2.0.so.0 \
```
Comment (jtojnar) [inline pkgs/development/libraries/gtkd/default.nix]: Maybe use `stdenv.lib.getLib` for greater accuracy?

File: pkgs/development/libraries/glib/default.nix [hunk 3]
```diff
@@ -57,7 +57,7 @@ stdenv.mkDerivation rec {
     ++ optional doCheck ./skip-timer-test.patch
     ++ [ ./schema-override-variable.patch ];
 
-  outputs = [ "out" "dev" "devdoc" ];
+  outputs = [ "bin" "out" "dev" "devdoc" ];
```
Comment (jtojnar) [inline pkgs/development/libraries/glib/default.nix]: Though I am not sure if the `bin` output should be the primary one. cc @vcunat

Comment (dezgeg) [reply pkgs/development/libraries/glib/default.nix]: Yes that's correct. The intention is anybody can write `${pkgs.foo}/bin/bar` in their systemd unit scripts or whatever and have it work.

Comment (vcunat) [reply pkgs/development/libraries/glib/default.nix]: Probably.  So far we've only broken the convention for `glibc` AFAIK.

Comment (vcunat) [reply pkgs/development/libraries/glib/default.nix]: but beware of all the code using `${glib}/lib/` and similar paths

**Comment by Mic92** [comment]
Thanks for finishing that!

**Pull request merged**